File: csharp/pattenrs/structural/Facade.cs
namespace Facade
{
    // Facade: provides a simplified interface to a large body of code.
    //
    // Motivation:
    // provide simple API for complicated underlying systems
    // allows to define one point of interaction between the client and complex system
    // reduce the number of dependencies between the client and a complex system
    class Main
    {
        public static void Run()
        {
            Console.WriteLine("\nFacade");

            // Facade - IDE

            // init facade
            var ide = new IDEFacade(new TextEditor(), new Compiler(), new Runtime(), new IDEConsole());
            // init client
            var client = new Developer(ide);
            // create app
            client.CreateApplication("Hello, world!");

            // Facade - Magic Square

            // init magic square generator
            var sg = new MagicSquareGenerator(new Generator(), new Splitter(), new Verifier());
            // generate magic square
            // var ms = sg.Generate(3);
        }
    }

    // -- Complex Underlying Systems

    // ITextEditor - represents text editor interface.
    public interface ITextEditor
    {
        public string WriteText(string text);
        public string SaveText();
    }

    // TextEditor - represents text editor.
    public class TextEditor : ITextEditor
    {
        private List<string> Stack = new List<string>();
        private string Current = "";

        public string WriteText(string text)
        {
            return Current += text;
        }

        public string SaveText()
        {
            Stack.Add(Current);
            Current = "";
            return Stack.Last();
        }
    }

    // ICompiler - represents Compiler interface.
    public interface ICompiler
    {
        public string Compile(string code);
    }

    // Compiler - represents code compiler.
    public class Compiler : ICompiler
    {
        public string Compile(string code)
        {
            return $"compiled({code})";
        }
    }

    // IRuntime - represents Runtime interface.
    public interface IRuntime
    {
        public string Execute(string bin);
    }

    // Runtime - represents runtime.
    public class Runtime : IRuntime
    {
        public string Execute(string bin)
        {
            return $"executed({bin})";
        }
    }

    // IConsole - represents console interface.
    public interface IConsole
    {
        public void Output(string result);
    }

    // IDEConsole - represents console.
    public class IDEConsole : IConsole
    {
        public void Output(string result)
        {
            Console.WriteLine($"output({result})");
        }
    }

    // -- Facade

    // IDE - represents IDE interface.
    public interface IDE
    {
        public void Run(string code);
    }

    // IDEFacade - represents IDE, which lavarage complex underlying systems to provide simple API.
    public class IDEFacade : IDE
    {
        protected ITextEditor editor;
        protected ICompiler compiler;
        protected IRuntime runtime;
        protected IConsole console;

        public IDEFacade(ITextEditor editor, ICompiler compiler, IRuntime runtime, IConsole console)
        {
            this.editor = editor;
            this.compiler = compiler;
            this.runtime = runtime;
            this.console = console;
        }

        public void Run(string appCode)
        {
            // write and save app code
            editor.WriteText(appCode);
            var code = editor.SaveText();

            // compile app code
            var bin = compiler.Compile(code);

            // execute binaries
            var result = runtime.Execute(bin);

            // output result
            console.Output(result);
        }
    }

    // -- Client

    // Developer - represent client
    public class Developer
    {
        private IDE tool;
        public Developer(IDE tool)
        {
            this.tool = tool;
        }

        public void CreateApplication(string appCode)
        {
            tool.Run(appCode);
        }
    }

    // -- Magic Square Example

    // -- Complex Underlying Systems

    // IGenerator - represents slice generator interface.
    public interface IGenerator
    {
        public List<int> Generate(int count);
    }

    // Generator - generates random slice of integers.
    public class Generator : IGenerator
    {
        private static readonly Random random = new Random();

        public List<int> Generate(int count)
        {
            return Enumerable.Range(0, count)
              .Select(_ => random.Next(1, 6))
              .ToList();
        }
    }

    // ISplitter - represents splitter interface.
    public interface ISplitter
    {
        public List<List<int>> Split(List<List<int>> array);
    }

    // Splitter - splits square matrix into individual parallel and diagonal pieces.
    public class Splitter : ISplitter
    {
        public List<List<int>> Split(List<List<int>> array)
        {
            var result = new List<List<int>>();

            var rowCount = array.Count;
            var colCount = array[0].Count;

            // get the rows
            for (int r = 0; r < rowCount; ++r)
            {
                var theRow = new List<int>();
                for (int c = 0; c < colCount; ++c)
                    theRow.Add(array[r][c]);
                result.Add(theRow);
            }

            // get the columns
            for (int c = 0; c < colCount; ++c)
            {
                var theCol = new List<int>();
                for (int r = 0; r < rowCount; ++r)
                    theCol.Add(array[r][c]);
                result.Add(theCol);
            }

            // now the diagonals
            var diag1 = new List<int>();
            var diag2 = new List<int>();
            for (int c = 0; c < colCount; ++c)
            {
                for (int r = 0; r < rowCount; ++r)
                {
                    if (c == r)
                        diag1.Add(array[r][c]);
                    var r2 = rowCount - r - 1;
                    if (c == r2)
                        diag2.Add(array[r][c]);
                }
            }

            result.Add(diag1);
            result.Add(diag2);

            return result;
        }
    }

    // IVerifier - represents verifier interface.
    public interface IVerifier
    {
        public bool Verify(List<List<int>> array);
    }

    // Verifier - verifies that it is indeed a maginc square.
    public class Verifier : IVerifier
    {
        public bool Verify(List<List<int>> array)
        {
            if (!array.Any()) return false;

            var expected = array.First().Sum();

            return array.All(t => t.Sum() == expected);
        }
    }

    // -- Facade

    // MagicSquareGenerator - represents facade, that leverage underlying systems to explode simple API.
    public class MagicSquareGenerator
    {
        private IGenerator generator;
        private ISplitter splitter;
        private IVerifier verifier;

        public MagicSquareGenerator(IGenerator generator, ISplitter splitter, IVerifier verifier)
        {
            this.generator = generator;
            this.splitter = splitter;
            this.verifier = verifier;
        }

        public List<List<int>> Generate(int size)
        {
            // generate data
            List<List<int>> lists = new List<List<int>>();
            for (int i = 0; i < size; i++)
            {
                lists.Add(generator.Generate(size));
            }

            // split data
            var split = splitter.Split(lists);

            // verify data
            if (!verifier.Verify(split))
            {
                return Generate(size);
            }

            return lists;
        }
    }
}

File: golang/patterns/structural/facade/facade.go
package facade

import (
	"fmt"
	"math/rand"
	"os"
	"time"
)

// Facade: provides a simplified interface to a large body of code.
//
// Motivation:
// provide simple API for complicated underlying systems
// allows to define one point of interaction between the client and complex system
// reduce the number of dependencies between the client and a complex system

func Facade() {
	fmt.Println("Facade")

	// Facade - IDE

	// init faacade
	ide := NewIDEFacade(NewTextEditor(), NewCompiler(), NewRuntime(), NewConsole())
	// init client
	client := NewDeveloper(ide)
	// create app
	client.CreateApplication("Hello, world!")

	// Facade - Magic Square Generator

	// init magic square generator
	_ = NewMagicSquareGenerator(NewGenerator(), NewSplitter(), NewVerifier())
	// generate magic square
	// ms := sg.GenerateSquare(3)
}

// -- Complex Underlying System

// TextEditorInterface - represents text editor interface.
type TextEditorInterface interface {
	WriteText(text string) string
	SaveText() string
}

// TextEditor - represents basic text editor.
type TextEditor struct {
	stack []string
	cur   string
}

// NewTextEditor - creates new instance of TextEditor.
func NewTextEditor() *TextEditor {
	return &TextEditor{[]string{}, ""}
}

// WriteText - write text into text editor.
func (te *TextEditor) WriteText(text string) string {
	te.cur += text
	return te.cur
}

// SaveText - save curent text state in text editor.
func (te *TextEditor) SaveText() string {
	te.stack = append(te.stack, te.cur)
	te.cur = ""
	return te.stack[len(te.stack)-1]
}

// CompilerInterface - represents compiler interface.
type CompilerInterface interface {
	Compile(code string) string
}

// Compiler - represents compiler.
type Compiler struct{}

// NewCompiler - creates new instance of Compiler.
func NewCompiler() *Compiler {
	return &Compiler{}
}

// Compile - compiles provided code.
func (c *Compiler) Compile(code string) string {
	return fmt.Sprintf("compiled(%s)", code)
}

// RuntimeInterface - represents runtime interface.
type RuntimeInterface interface {
	Execute(bin string) string
}

// Runtime - represents application runtime.
type Runtime struct{}

// NewRuntime - creates new instance of Runtime.
func NewRuntime() *Runtime {
	return &Runtime{}
}

// Execute - executes provided binaries.
func (r *Runtime) Execute(bin string) string {
	return fmt.Sprintf("executed(%s)", bin)
}

// ConsoleInterface - represents console interface.
type ConsoleInterface interface {
	Output(result string)
}

// Console - represents console.
type Console struct{}

// NewConsole - creates new instance of Console.
func NewConsole() *Console {
	return &Console{}
}

// Output - outputs execution result into standart output.
func (cl *Console) Output(result string) {
	fmt.Fprintf(os.Stdout, "output(%s)\n", result)
}

// -- Facade

// IDEInterface - represents IDE interface.
type IDEInterface interface {
	Run(code string)
}

// IDEFacade - represents IDE, which lavarage complex underlying systems to provide simple API.
type IDEFacade struct {
	TextEditorInterface
	CompilerInterface
	RuntimeInterface
	ConsoleInterface
}

// NewIDEFacade - creates new instance of IDEFacade.
func NewIDEFacade(
	te TextEditorInterface,
	cm CompilerInterface,
	rn RuntimeInterface,
	cl ConsoleInterface,
) *IDEFacade {
	return &IDEFacade{te, cm, rn, cl}
}

// Run - runs application code.
func (ide *IDEFacade) Run(text string) {
	ide.WriteText(text)
	code := ide.SaveText()
	bin := ide.Compile(code)
	out := ide.Execute(bin)
	ide.Output(out)
}

// -- Client

// Developer - represents client.
type Developer struct {
	tool IDEInterface
}

// NewDeveloper - creates new instance of developer.
func NewDeveloper(tool IDEInterface) *Developer {
	return &Developer{tool}
}

// CreateApplication - creates application.
func (d *Developer) CreateApplication(appCode string) {
	d.tool.Run(appCode)
}

// -- Magic Square Generator

// -- Complex Underlying System

// GeneratorInterface - represents generator interface.
type GeneratorInterface interface {
	GenerateSlice(n int) []int
}

// Generator - represents slice generator.
type Generator struct{}

// NewGenerator - creates new instance of Generator.
func NewGenerator() *Generator {
	return &Generator{}
}

// GenerateSlice - generates slice of n size.
func (g *Generator) GenerateSlice(n int) []int {
	rand.Seed(time.Now().UnixNano())
	min := 1
	max := 6

	sl := make([]int, n)
	for i := 0; i < n; i++ {
		sl[i] = rand.Intn(max-min) + min
	}
	return sl
}

// SplitterInterface - represents splitter interface.
type SplitterInterface interface {
	Split(inp [][]int) [][]int
}

// Splitter - represents splitter.
type Splitter struct{}

// NewSplitter - creates new instance of Splitter.
func NewSplitter() *Splitter {
	return &Splitter{}
}

// Split - splits square matrix into individual parallel and diagonal pieces.
func (s *Splitter) Split(inp [][]int) [][]int {
	rows := len(inp)
	cols := len(inp[0])
	res := make([][]int, 0)

	// get rows
	for i := 0; i < rows; i++ {
		row := make([]int, 0)
		for j := 0; j < cols; j++ {
			row = append(row, inp[i][j])
		}
		res = append(res, row)
	}

	// get cols
	for i := 0; i < cols; i++ {
		col := make([]int, 0)
		for j := 0; j < rows; j++ {
			col = append(col, inp[j][i])
		}
		res = append(res, col)
	}

	// get diagonals
	dia1 := make([]int, 0)
	dia2 := make([]int, 0)
	for i := 0; i < cols; i++ {
		for j := 0; j < rows; j++ {
			if i == j {
				dia1 = append(dia1, inp[j][i])
			}
			k := rows - j - 1
			if i == k {
				dia2 = append(dia2, inp[j][i])
			}
		}
	}
	res = append(res, dia1)
	res = append(res, dia2)

	fmt.Println(res)
	return res
}

// VerifierInterface - represents verifier interface.
type VerifierInterface interface {
	Sum(sl []int) int
	Verify(res [][]int) bool
}

// Verifier - represents verifier.
type Verifier struct{}

// NewVerifier - creates new instance of Verifier.
func NewVerifier() *Verifier {
	return &Verifier{}
}

// Verify - verifies that provided slices have equal sum each.
func (v *Verifier) Verify(res [][]int) bool {
	sum := v.Sum(res[0])
	// fmt.Println(sum)
	for _, sl := range res {
		// fmt.Println(res)
		if sum != v.Sum(sl) {
			return false
		}
	}

	return true
}

// Sum - sums slice elements.
func (v *Verifier) Sum(sl []int) int {
	sum := 0
	for _, n := range sl {
		sum += n
	}
	return sum
}

// -- Facade

// MagicSquareGenerator - represents magic square generator facade.
type MagicSquareGenerator struct {
	GeneratorInterface
	SplitterInterface
	VerifierInterface
}

// NewMagicSquareGenerator - creates new instance of MagicSquareGenerator.
func NewMagicSquareGenerator(
	generator GeneratorInterface,
	splitter SplitterInterface,
	verifier VerifierInterface,
) *MagicSquareGenerator {
	return &MagicSquareGenerator{generator, splitter, verifier}
}

// GenerateSquare - generates magic square.
func (msg *MagicSquareGenerator) GenerateSquare(size int) [][]int {

	// generate square
	sq := make([][]int, 0)
	for i := 0; i < size; i++ {
		sq = append(sq, msg.GenerateSlice(size))
	}

	// split square into pieces
	res := msg.Split(sq)

	// verify that magic square is valid
	if !msg.Verify(res) {
		// retry if square is invalid
		return msg.GenerateSquare(size)
	}

	return sq
}
